tests/distributions/test_distributions.py

```diff
@@ -27,34 +27,36 @@ def _log_prob_shape(dist, x_size=torch.Size(), **dist_params):
 
 def test_batch_log_prob(dist):
     if dist.scipy_arg_fn is None:
-        pytest.skip('{}.log_pdf has no scipy equivalent'.format(dist.pyro_dist_obj.__name__))
-    d = dist.pyro_dist
+        pytest.skip('{}.log_pdf has no scipy equivalent'.format(dist.pyro_dist_class.__name__))
     for idx in dist.get_batch_data_indices():
         dist_params = dist.get_dist_params(idx)
+        d = dist.pyro_dist_class(**dist_params)
         test_data = dist.get_test_data(idx)
-        log_prob_sum_pyro = _unwrap_variable(d.log_prob(test_data, **dist_params)).sum()
+        log_prob_sum_pyro = _unwrap_variable(d.log_prob(test_data)).sum()
         log_prob_sum_np = np.sum(dist.get_scipy_batch_logpdf(-1))
         assert_equal(log_prob_sum_pyro, log_prob_sum_np)
 
 
 def test_sample_shape(dist):
     for idx in range(dist.get_num_test_data()):
+        if dist.pyro_dist is None:
+            continue  # stateful distributions cannot implement a function interface
         dist_params = dist.get_dist_params(idx)
         x_func = dist.pyro_dist.sample(**dist_params)
-        x_obj = dist.pyro_dist_obj(**dist_params).sample()
+        x_obj = dist.pyro_dist_class(**dist_params).sample()
         assert_equal(x_obj.size(), x_func.size())
 
 
 def test_batch_log_prob_shape(dist):
-    d = dist.pyro_dist
     for idx in range(dist.get_num_test_data()):
         dist_params = dist.get_dist_params(idx)
+        d = dist.pyro_dist_class(**dist_params)
         x = dist.get_test_data(idx)
         with xfail_if_not_implemented():
             # Get log_prob shape after broadcasting.
-            expected_shape = _log_prob_shape(d, x.size(), **dist_params)
-            log_p_func = d.log_prob(x, **dist_params)
-            log_p_obj = dist.pyro_dist_obj(**dist_params).log_prob(x)
+            expected_shape = _log_prob_shape(d, x.size())
+            log_p_func = d.log_prob(x)
```

Comment:
This is the same as `log_p_obj` now. While checking for consistency between RandomPrimitive and Distribution instances isn't that important now, it ensures that we are calling the RandomPrimitive code path for all major methods without throwing exceptions. Can we just do 
```
d=dist.pyro_dist_class(**dist_params) if dist.pyro_dist is None else dist.pyro_dist
```
, or remove `log_p_func` and redundant checks?

Reply:
Done.